From 4fd9295b34053c807ba1848533b55ad2b345e9b7 Mon Sep 17 00:00:00 2001
From: Ioana Radulescu <ruxandra.radulescu@nxp.com>
Date: Fri, 2 Feb 2018 13:51:09 +0200
Subject: [PATCH 386/767] staging: fsl-dpaa2/eth: Guard call to
 bpf_prog_run_xdp

XDP programs must be run with the RCU lock taken.

Signed-off-by: Ioana Radulescu <ruxandra.radulescu@nxp.com>
[Xulin: Original patch taken from NXP LSDK-18.09.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c
index 506da8e027c6..f1614d2030cc 100644
--- a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c
+++ b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c
@@ -319,6 +319,7 @@ static u32 dpaa2_eth_run_xdp(struct dpaa2_eth_priv *priv,
 	/* Allow the XDP program to use the specially reserved headroom */
 	xdp.data_hard_start = xdp.data - XDP_PACKET_HEADROOM;
 
+	rcu_read_lock();
 	xdp_act = bpf_prog_run_xdp(xdp_prog, &xdp);
 
 	/* xdp.data pointer may have changed */
@@ -361,6 +362,8 @@ static u32 dpaa2_eth_run_xdp(struct dpaa2_eth_priv *priv,
 		percpu_stats->rx_bytes += dpaa2_fd_get_len(fd);
 	}
 
+	rcu_read_unlock();
+
 	return xdp_act;
 }
 
-- 
2.17.0

